### PR TITLE
Define return values for Ultrasonic error cases

### PR DESCRIPTION
**Pull request by baylessj**
In response to Issue #32 , this pull request defines two types of error values for the Ultrasonic sensor, one for when the ultrasonic sensor is polled while it is pinging and waiting for a value (ULTRASONIC_TIMEOUT), and one when the ultrasonic is reading a value that is outside of its range (ULTRASONIC_BEYOND_RANGE).

### CODE REVIEW

**Comment by baylessj** [comment]
The changes were tested by polling a single ultrasonic in the operator control loop at the default time interval of 20 ms. 

The conversation in Issue #32 seemed to indicate that the sensor would return ULTRASONIC_TIMEOUT every 50ms or so, but this value was returned very rarely, roughly once per 10-15 seconds of use. ULTRASONIC_BEYOND_RANGE values were improperly returned in situations where there was an object within the sensor's range much more often than the ULTRASONIC_TIMEOUT values were returned, at roughly once per second. This improper return value might be the result of testing the sensor with surfaces that are do not reflect the sound perfectly, thus causing the sensor to fail to detect the objects.

This change made it much easier to identify the difference in error values and also recognize the source of error in ultrasonic data when debugging by printing the values to the terminal.

**Comment by HotelCalifornia** [comment]
> This change made it much easier to identify the difference in error values and also recognize the source of error in ultrasonic data when debugging by printing the values to the terminal.
Do you have an example of the outputs for each case?

<sup>if only to satisfy my curiosity 😉</sup>

**Comment by baylessj** [comment]
```
distance: -2
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: 26
distance: 26
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: -1
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 25
distance: 26
distance: 26
distance: 26
distance: 26
distance: -1
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: -1
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 26
distance: 27
distance: 27
distance: 27
distance: 27
distance: 27
distance: 27
distance: 27
distance: 29
distance: 29
distance: 29
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -2
distance: -1
distance: -1
distance: -2
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -2
distance: -1
distance: -1
distance: -2
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
distance: -1
distance: -2
distance: -1
distance: -1
distance: -1
```

Well it would appear that my recollection of my previous data collection was incorrect (or the ultrasonic has started behaving differently since then, but that's unlikely), and the timeout values were returned every now and then, even when a flat surface was placed in front of it.

The test was conducted in the same manner as before, with the ultrasonic returning values every 20ms in the operator control loop. There was no surface in the ultrasonic's range at first, then a flat surface was introduced (the lid to the box for the Battleship board game, fwiw), and then the flat surface was removed. One particularly interesting thing to note is that the ping timed out more often than it returned the ULTRASONIC_BEYOND_RANGE value, but it's still clear that at that point in time it was garbage data because the Ultrasonic sensor was out of range.

File: include/API.h [hunk 1]
```diff
@@ -654,9 +659,10 @@ typedef void * Ultrasonic;
 /**
  * Gets the current ultrasonic sensor value in centimeters.
  *
- * If no object was found, zero is returned. If the ultrasonic sensor was never started, the
- * return value is undefined. Round and fluffy objects can cause inaccurate values to be
- * returned.
+ * If no object was found, -2 (ULTRASONIC_BEYOND_RANGE) is returned. If the ultrasonic sensor is
+ * polled while it is pinging and waiting for a response, -1 (ULTRASONIC_TIMEOUT) is returned.
```
Comment (edjubuh) [inline include/API.h]: This documentation needs to be updated

**Comment by baylessj** [comment]
Updated the pros-docs documentation in commit purduesigbots/pros-docs@a5cb8854d452afa3fd9fa32037c417664a8317fc

**Comment by baylessj** [comment]
Previous commit to master was reverted, the changes can now be found in purduesigbots/pros-docs#26

**Comment by Octogonapus** [comment]
Inside of `fireUltrasonic` why do we write `ULTRA_BAD_RESPONSE` to `value` before the ping? The best guess we can make for the next ping is the value of the previous ping, not an invalid response. What about not writing to `value` at all before the ping?

**Comment by baylessj** [comment]
That point of debate was covered more thoroughly in Issue #32 , where the case of an object moving from the sensor's range to beyond its range was considered. If a default value is set, then the user would be aware that the object had moved out of the ultrasonic's range by the first ping that was not properly returned, but if the old value is kept, then it would appear as though the object never moved out of the ultrasonic's range. 

Keeping the old value removes the need for the user to filter the ultrasonic's returned values, but at the cost of not providing all of the available information about what the ultrasonic is observing.

**Comment by Octogonapus** [comment]
Okay, if we are going with that then it all looks good 👍

**Comment by HotelCalifornia** [comment]
Just want to get another test in, with a new cortex and a different ultrasonic sensor so we can confirm this will work properly

**Comment by Octogonapus** [comment]
Tested using this code: https://hastebin.com/amusegonoc.cpp
Got this output: https://hastebin.com/lupecelaji.scala

Everything looks as expected. I tested the normal range and then moved into what should be outside of the range, and the return values look good.

**Comment by HotelCalifornia** [comment]
@Octogonapus your test code looks like it's using an older revision of this PR, before we moved from two defined values to just one (`ULTRA_BAD_RESPONSE`)

**Comment by Octogonapus** [comment]
@HotelCalifornia That's correct, my bad. I tested again using the correct version and got similar results (everything still looks good).

**Pull request merged**